sjsonnet/src/sjsonnet/YamlRenderer.scala

```diff
@@ -155,9 +174,23 @@ class YamlRenderer(_out: StringWriter = new java.io.StringWriter(), indentArrayI
   }
 }
 object YamlRenderer{
-  val newlinePattern = Pattern.compile("\n")
+  val newlinePattern: Pattern = Pattern.compile("\n")
+  private val safeYamlKeyPattern = Pattern.compile("^[a-zA-Z0-9/._-]+$")
+  private val yamlReserved = Set("true", "false", "null", "yes", "no", "on", "off", "y", "n", ".nan",
+    "+.inf", "-.inf", ".inf", "null", "-", "---", "''")
+  private val yamlTimestampPattern = Pattern.compile("^(?:[0-9]*-){2}[0-9]*$", Pattern.CASE_INSENSITIVE)
+  private val yamlBinaryPattern = Pattern.compile("^[-+]?0b[0-1_]+$")
+  private val yamlHexPattern = Pattern.compile("[-+]?0x[0-9a-fA-F_]+")
+
+  private def isSafeBareKey(k: String) = {
+    val l = k.toLowerCase
+    !yamlReserved.contains(l) && safeYamlKeyPattern.matcher(k).matches() &&
+      !yamlTimestampPattern.matcher(k).matches() && !yamlBinaryPattern.matcher(k).matches() &&
+      !yamlHexPattern.matcher(k).matches() && (Try(l.replace("_", "").toLong).isFailure
+      && Try(l.replace("_", "").toDouble).isFailure)
```

Comment:
Since most keys are valid, this will on average result in throwing and catching two exceptions per key, which is going to be disproportionately expensive compared to everything else in this method:

![image](https://github.com/user-attachments/assets/632e0bba-6a24-469b-a996-4e1fb689dbf5)

I think it might be possible to use an additional regex to check if something is an integer or float.

Reply:
In `jsonnet`, this is implemented at https://github.com/google/jsonnet/blob/913281d203578bb394995bacc792f2576371e06c/stdlib/std.jsonnet#L1119-L1134 . 

---

Because their stdlib is implemented in jsonnet itself and they lack regex, some of their implementation is a bit weird. For example:

```
      /* Check for integers.  Keys that meet all of the following:
           - all characters match [0-9_\-]
           - has at most 1 dash
         are considered integers.
      */
      else if onlyChars(intChars, keySetLc)
              && std.length(std.findSubstr('-', key)) < 2 then
```

I imagine that we don't intend to be _that_ jsonnet-compliant, though, since it doesn't look like even they have test cases for corner cases like `1-2-3`, etc.

---

I think we can pass all of the tests with

```scala
  val newlinePattern: Pattern = Pattern.compile("\n")
  private val safeYamlKeyPattern = Pattern.compile("^[a-zA-Z0-9/._-]+$")
  private val yamlReserved = Set("true", "false", "null", "yes", "no", "on", "off", "y", "n", ".nan",
    "+.inf", "-.inf", ".inf", "null", "-", "---", "''")
  private val yamlTimestampPattern = Pattern.compile("^(?:[0-9]*-){2}[0-9]*$")
  private val yamlBinaryPattern = Pattern.compile("^[-+]?0b[0-1_]+$")
  private val yamlHexPattern = Pattern.compile("[-+]?0x[0-9a-fA-F_]+")
  private val yamlFloatPattern = Pattern.compile(
    "^-?(?>[0-9_]*)*(?>\\.[0-9_]*)?(?>e[-+][0-9_]+)?$"
  )
  private val yamlIntPattern = Pattern.compile("^[-+]?[0-9_]+$")

  private def isSafeBareKey(k: String) = {
    val l = k.toLowerCase
    !yamlReserved.contains(l) &&
      safeYamlKeyPattern.matcher(k).matches() &&
      !yamlTimestampPattern.matcher(l).matches() &&
      !yamlBinaryPattern.matcher(k).matches() &&
      !yamlHexPattern.matcher(k).matches() &&
      !yamlFloatPattern.matcher(l).matches() &&
      !yamlIntPattern.matcher(l).matches()
  }
```

where I've gotten a bit fancy with the float pattern with the use of atomic groups to avoid backtracking. I've also chosen to handle the underscores in the regex rather than using `str.replace()` to strip them.

Reply:
I kind of followed what go-sjsonnet is doing, but working around java's inconsistencies with _ separator (they are invalid in parseInt() functions.)
Let's use your version. Seems to work